test/test_ops.py

```diff
@@ -244,34 +244,47 @@ def test_noncontiguous_samples(self, device, dtype, op):
             self.assertEqual(actual, expected)
 
             # validates backward
-            # NOTE: only handles single tensor outputs and the first tensor
-            #   of ops that output a sequence
 
             # Short-circuits if the op doesn't support grad in this device x dtype
             if not test_grad:
                 continue
 
+            def assertEqualGradients(expected, actual, i=None):
```

Comment:
Did you walk through an example locally to ensure this is doing what you expect? 

These tests are so complicated I always have to do that to feel comfortable they're testing anything

Reply:
Yep, and I did not find anything weird.

They also failed when the generation of the inputs was not done in an independent manner (i.e. some inputs were views of others) which is reassuring.

Reply:
I have heavily simplified this test, and now it should be easier to reason about it. Could you have another look?

Reply:
The updated design looks great!